### PR TITLE
Simple entity editor

### PR DESCRIPTION
**Pull request by Karaulov**
I tried make entity editor with amxx, but **public pfn_keyvalue(ent)** do make impossible to edit classname of enities. 

This is very easy entity editor for any map. I don't tried add new entities, but possible need add them to end of file. File .ent created while map loading. Path /maps/mapname.ent .

Example video "how it works" on cs_assault map: https://youtu.be/Rkxcapg4rOU
_Sorry for very bad video fps, my pc 10 years old )))_

### CODE REVIEW

File: rehlds/engine/sv_main.cpp [hunk 1]
```diff
@@ -6126,6 +6127,48 @@ int SV_SpawnServer(qboolean bIsDemo, char *server, char *startspot)
...
+			if ( !f )
+			{
+				ED_LoadFromFile(g_psv.worldmodel->entities);
+				return;
+			}
+			FS_Write( g_psv.worldmodel->entities, strlen( g_psv.worldmodel->entities ), 1, f );
+			FS_Close( f );
+		}
+
+		if ( FS_FileExists( va( "%s/%s.ent", "maps/", g_psv.name ) ) )
```
Comment (LevShisterov) [inline rehlds/engine/sv_main.cpp]: File existance double check. Replace with `else`.
Also, this project uses a different formatting for parentnesis (no spaces inside).

File: rehlds/engine/sv_main.cpp [hunk 2]
```diff
@@ -6126,6 +6127,48 @@ int SV_SpawnServer(qboolean bIsDemo, char *server, char *startspot)
...
+		if ( !FS_FileExists( va( "%s/%s.ent", "maps/", g_psv.name ) ) )
+		{
+			FILE *f = FS_Open( va( "%s/%s.ent", "maps/", g_psv.name ), "wb" );
+			if ( !f )
+			{
+				ED_LoadFromFile(g_psv.worldmodel->entities);
+				return;
+			}
+			FS_Write( g_psv.worldmodel->entities, strlen( g_psv.worldmodel->entities ), 1, f );
+			FS_Close( f );
```
Comment (LevShisterov) [inline rehlds/engine/sv_main.cpp]: Change to not "write then read file". Just write and then use the value you already have at hands and return.

File: rehlds/engine/sv_main.cpp (hunk 1 above)
Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: This check is need.

File: rehlds/engine/sv_main.cpp (hunk 2 above)
Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: Sorry but i can't understand what you mean.

1. Create file if not exist and write default entities.
2. If file already exist, use entities from this file.

File: rehlds/engine/sv_main.cpp (hunk 1 above)
Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: Or not. Yes it can be editet) First file is default entities.

File: rehlds/engine/sv_main.cpp [hunk 3]
```diff
@@ -6126,6 +6127,47 @@ int SV_SpawnServer(qboolean bIsDemo, char *server, char *startspot)
...
+				return;
+			}
+			unsigned int nFileSize = FS_Size(f);
+			char *pszInputStream;
+			int nBytesRead;
+			pszInputStream = (char *)Mem_ZeroMalloc(nFileSize + 1);
+			nBytesRead = FS_Read(pszInputStream, nFileSize, 1, f);
+			auto oldval = g_psv.worldmodel->entities;
+			g_psv.worldmodel->entities = pszInputStream;
+			ED_LoadFromFile(g_psv.worldmodel->entities);
```
Comment (s1lentq) [inline rehlds/engine/sv_main.cpp]: Just use ED_LoadFromFile(pszInputStream)
In my opinion there is not necessary to use g_psv.worldmodel->entities

Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: Sorry i don't know where used g_psv.worldmodel->entities while ED_LoadFromFIle called and replace it by new buffer. Okay i can edit it.

Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: Fixed?

Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: I fix it 
```suggestion
			ED_LoadFromFile
```

File: rehlds/engine/sv_main.cpp [hunk 4]
```diff
@@ -6126,6 +6127,42 @@ int SV_SpawnServer(qboolean bIsDemo, char *server, char *startspot)
...
 {
+#ifdef REHLDS_FIXES
+	if (sv_entityeditor.value > 0.0f)
+	{
+		Con_Printf("sv_entityeditor = 1\nInitialized entity editor.\nEntity path: maps/%s.ent\n", g_psv.name);
+
+		if (!FS_FileExists(va("%s/%s.ent", "maps/", g_psv.name)))
+		{
+			FILE *f = FS_Open(va("%s/%s.ent", "maps/", g_psv.name), "wb");
+			if (!f)
```
Comment (LevShisterov) [inline rehlds/engine/sv_main.cpp]: Invert `if` and you will fallback to default loading code.

File: rehlds/engine/sv_main.cpp (hunk 3 above)
Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: I fix it

File: rehlds/engine/sv_main.cpp [hunk 5]
```diff
@@ -6126,6 +6127,42 @@ int SV_SpawnServer(qboolean bIsDemo, char *server, char *startspot)
...
+				ED_LoadFromFile(g_psv.worldmodel->entities);
+				return;
+			}
+			FS_Write(g_psv.worldmodel->entities, strlen(g_psv.worldmodel->entities), 1, f);
+			FS_Close(f);
+		}
+		else
+		{
+			FILE *f = FS_Open(va("%s/%s.ent", "maps/", g_psv.name), "rb");
+			if (!f)
```
Comment (LevShisterov) [inline rehlds/engine/sv_main.cpp]: Same here.

File: rehlds/engine/sv_main.cpp (hunk 4 above)
Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: Done

File: rehlds/engine/sv_main.cpp (hunk 5 above)
Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: Done

**Comment by LevShisterov** [comment]
Cool. Thanks! A bit more changes: I suggest to change "sv_entityeditor" into "sv_use_entity_file" and a bit text corrections where editor is mensioned. Because word "editor" actually could mislead that there is some kind of editor inside. And add new cvar with description to README.md file.

**Comment by Karaulov** [comment]
> Cool. Thanks! A bit more changes: I suggest to change "sv_entityeditor" into "sv_use_entity_file" and a bit text corrections where editor is mensioned. Because word "editor" actually could mislead that there is some kind of editor inside. And add new cvar with description to README.md file.

I can edit README.md and add it to this pull request?? I don't understand how it works in github :)

**Comment by Karaulov** [comment]
Ok. I rename cvar to sv_use_entity_file and add to README.md (but I am not sure in English language, if necessary correct me)

**Comment by LevShisterov** [comment]
@Karaulov please check that it stil working at intended. Edited via Github.

**Comment by Karaulov** [comment]
@LevShisterov yes, it works.

File: rehlds/engine/sv_main.cpp [hunk 6]
```diff
@@ -6126,6 +6127,49 @@ int SV_SpawnServer(qboolean bIsDemo, char *server, char *startspot)
...
+#ifdef REHLDS_FIXES
+	if (sv_use_entity_file.value > 0.0f)
+	{
+		char name[MAX_PATH];
+		Q_snprintf(name, sizeof(name), "maps/%s.ent", g_psv.name);
+
+		if (!FS_FileExists(name))
+		{
+			FILE *f = FS_Open(name, "wb");
+			if (f != NULL)
```
Comment (WPMGPRoSToTeMa) [inline rehlds/engine/sv_main.cpp]: `nullptr` please.

Comment (WPMGPRoSToTeMa) [inline rehlds/engine/sv_main.cpp]: `nullptr` please.

File: rehlds/engine/sv_main.cpp [hunk 7]
```diff
@@ -6126,6 +6127,49 @@ int SV_SpawnServer(qboolean bIsDemo, char *server, char *startspot)
...
+		else
+		{
+			FILE *f = FS_Open(name, "rb");
+			if (f != NULL)
+			{
+				Con_Printf("Using custom entity file: %s\n", name);
+
+				unsigned int nFileSize = FS_Size(f);
+				char *pszInputStream = (char *)Mem_Malloc(nFileSize + 1);
+				if (!pszInputStream)
```
Comment (WPMGPRoSToTeMa) [inline rehlds/engine/sv_main.cpp]: Check against `nullptr` for style consistency with the code above.

File: rehlds/engine/sv_main.cpp [hunk 8]
```diff
@@ -6126,6 +6127,49 @@ int SV_SpawnServer(qboolean bIsDemo, char *server, char *startspot)
...
+			}
+		}
+		else
+		{
+			FILE *f = FS_Open(name, "rb");
+			if (f != NULL)
+			{
+				Con_Printf("Using custom entity file: %s\n", name);
+
+				unsigned int nFileSize = FS_Size(f);
```
Comment (WPMGPRoSToTeMa) [inline rehlds/engine/sv_main.cpp]: Should we avoid Hungarian notation?

File: rehlds/engine/sv_main.cpp (hunk 7 above)
Comment (LevShisterov) [reply rehlds/engine/sv_main.cpp]: https://github.com/dreamstalker/rehlds/blob/master/rehlds/engine/sv_main.cpp#L2598

File: rehlds/engine/sv_main.cpp (hunk 6 above)
Comment (LevShisterov) [reply rehlds/engine/sv_main.cpp]: A lot of the code uses NULL. Probably it is better to replace in the whole project first.

File: rehlds/engine/sv_main.cpp (hunk 8 above)
Comment (LevShisterov) [reply rehlds/engine/sv_main.cpp]: https://github.com/dreamstalker/rehlds/blob/master/rehlds/engine/sv_main.cpp#L2591
The same as with nullptr but look more as fighting with mills.

File: rehlds/engine/sv_main.cpp (hunk 6 above)
Comment (Karaulov) [reply rehlds/engine/sv_main.cpp]: @WPMGPRoSToTeMa @LevShisterov i just replace it to **if (f)**  http://cpp.sh/4jifk

File: rehlds/engine/sv_main.cpp (hunk 7 above)
Comment (WPMGPRoSToTeMa) [reply rehlds/engine/sv_main.cpp]: @LevShisterov both ways are acceptable. I just want to see exactly one way used (consistency) in these changes, doesn't matter which way he would choose.

So, the current code is OK now.

File: README.md [hunk 9]
```diff
@@ -55,6 +55,7 @@ Bugfixed version of rehlds contains an additional cvars:
 <li>sv_rehlds_userinfo_transmitted_fields // Userinfo fields only with these keys will be transmitted to clients via network. If not set then all fields will be transmitted (except prefixed with underscore). Each key must be prefixed by backslash, for example "\name\model\*sid\*hltv\bottomcolor\topcolor". See [wiki](https://github.com/dreamstalker/rehlds/wiki/Userinfo-keys) to collect sufficient set of keys for your server. Default: ""
 <li>sv_rehlds_attachedentities_playeranimationspeed_fix // Fixes bug with gait animation speed increase when player has some attached entities (aiments). Can cause animation lags when cl_updaterate is low. Default: 0
 <li>sv_rehlds_maxclients_from_single_ip // Limit number of connections from the single ip address. Default: 5
+<li>sv_use_entity_file // Use custom entity file for a map. Path to an entity file will be '/maps/[map name].ent'. Default: 0
```
Comment (WPMGPRoSToTeMa) [inline README.md]: Use backticks (`) instead of ' please.

Comment (Karaulov) [reply README.md]: Ok, but in other cases used '

Comment (WPMGPRoSToTeMa) [reply README.md]: Hmm, it didn't help, looks like you should use `<pre></pre>`.

Comment (Karaulov) [reply README.md]: I do not understand what you want ?!

Comment (WPMGPRoSToTeMa) [reply README.md]: @Karaulov code wrap 
image

Comment (Karaulov) [reply README.md]: It strange, i see it like: 
XMpprx5 1

Comment (Karaulov) [reply README.md]: I don't know what you mean, you can fix it yourself?

**Pull request merged**